### PR TITLE
Handle Network Peering in `apinetlet` and `metalnetlet`

### LINKED ISSUE #251: Implement Peering Translation in `apinetlet` and `metalnetlet`
## Overview
This issue aims to introduce peering translation logic within the `apinetlet` component to ensure seamless interoperability between the `networking.ironcore.dev` and `core.apinet.ironcore.dev` groups. The objective is to enable the `apinet` Network to automatically recognize and reflect peering relationships defined in the `networking.ironcore.dev` Network objects. Additionally, we will outline a conceptual approach for mapping these relationships to the underlying `metalnet` project via `metalnetlet`.

## Problem Statement
Currently, the `apinet` Network lacks the capability to automatically recognize or establish peering relationships defined in `networking.ironcore.dev` Network objects. This gap hinders seamless network interconnectivity and management across different layers of our networking stack.

### Scope
1. **`apinetlet` Peering Translation:** Implement logic to translate peering information from `networking.ironcore.dev` Network objects to `core.apinet.ironcore.dev` Network objects, ensuring that peering details are accurately reflected and maintained.
2. **Concept for `metalnet` Mapping:** Develop a high-level concept for mapping these peering relationships further down to the `metalnet` project through `metalnetlet`, ensuring end-to-end network connectivity and peering integrity.

### `apinetlet` Implementation
- **Peering Information Translation:** Automatically translate and synchronize peering information from `networking.ironcore.dev` Network objects to `core.apinet.ironcore.dev` Network objects.
  - This includes reflecting peer network names, namespaces, and unique identifiers.
- **Status Updates:** Ensure that the status of `core.apinet.ironcore.dev` Network objects is updated to reflect the state of peering relationships accurately.
- **Event Handling:** Implement event listeners to handle updates or modifications to peering information in `networking.ironcore.dev` Network objects and propagate changes accordingly.

### Concept for `metalnet` Mapping
- Develop a conceptual approach for translating and applying peering information from `apinetlet` to `metalnetlet`, ensuring compatibility and integrity in network configuration and peering at the `metalnet` level.
- Outline the necessary transformations or mappings required to align `networking.ironcore.dev` and `core.apinet.ironcore.dev` peering configurations with `metalnet` project expectations.

## Acceptance Criteria
- **Peering Translation Logic:** `apinetlet` correctly translates and synchronizes peering information from `networking.ironcore.dev` Network objects to `core.apinet.ironcore.dev` Network objects without manual intervention.
- **Status Accuracy:** The status of `core.apinet.ironcore.dev` Network objects accurately reflects the peering relationships and is dynamically updated in response to changes.
- **Event Handling:** `apinetlet` effectively responds to events related to peering information changes, ensuring timely and accurate propagation of updates.
- **Conceptual Mapping to `metalnet`:** A clear and actionable concept for mapping peering information to the `metalnet` project through `metalnetlet` is developed and documented.

## Next Steps
1. **Design Phase:** Detail the design of the peering translation logic and the conceptual mapping to `metalnet`, including data flow diagrams and API specifications.
2. **Implementation Phase:** Based on the approved designs, proceed with the implementation of the peering translation logic in `apinetlet` and the conceptual approach for `metalnet` mapping.
3. **Testing & Validation:** Conduct thorough testing to ensure that the peering translation and status updates function as expected across `networking.ironcore.dev` and `core.apinet.ironcore.dev` Network objects. Validate the conceptual mapping with a prototype or simulation, if feasible.

---
Please provide feedback on the proposed approach or any additional considerations that should be taken into account.

### PR DESCRIPTION
**Pull request by kasabe28**
# Proposed Changes

- Modify network controller in `apinetlet` to handle peerings
- Add tests for network peering in `apinetlet`
- Modify `metalnetlet` network controller to add `peeredIDs`
- Modify test for `metalnet` network

Fixes #251

### CODE REVIEW

File: apinetlet/controllers/network_peering_controller.go [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+	ctx context.Context,
+	log logr.Logger,
+	network *networkingv1alpha1.Network,
+	peeringNames []string,
+) error {
+	base := network.DeepCopy()
+	newStatusPeerings := make([]networkingv1alpha1.NetworkPeeringStatus, 0, len(peeringNames))
+	for _, name := range peeringNames {
+		newStatusPeerings = append(newStatusPeerings, networkingv1alpha1.NetworkPeeringStatus{
+			Name: name,
```
Comment (balpert89) [inline apinetlet/controllers/network_peering_controller.go:108]: add a short TODO to indicate this needs to be extended once the PR in ironcore for status extension has been merged

File: apinetlet/controllers/network_peering_controller.go [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+	if targetNetworkNamespace == "" {
+		targetNetworkNamespace = network.Namespace
+	}
+	targetNetworkKey := client.ObjectKey{Namespace: targetNetworkNamespace, Name: targetNetworkRef.Name}
+	log = log.WithValues("TargetNetworkKey", targetNetworkKey)
+
+	log.V(2).Info("Getting target network")
+	if err := r.Get(ctx, targetNetworkKey, targetNetwork); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return networkingv1alpha1.NetworkPeeringClaimRef{}, "", "", fmt.Errorf("error getting network %s: %w", targetNetworkKey, err)
```
Comment (balpert89) [inline apinetlet/controllers/network_peering_controller.go:187]: extend the error that this is the target / peering partner network

File: apinetlet/controllers/network_peering_controller.go [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+	targetNetwork := &networkingv1alpha1.Network{}
+	targetNetworkRef := peering.NetworkRef
+	targetNetworkNamespace := targetNetworkRef.Namespace
+	if targetNetworkNamespace == "" {
+		targetNetworkNamespace = network.Namespace
+	}
+	targetNetworkKey := client.ObjectKey{Namespace: targetNetworkNamespace, Name: targetNetworkRef.Name}
+	log = log.WithValues("TargetNetworkKey", targetNetworkKey)
+
+	log.V(2).Info("Getting target network")
```
Comment (balpert89) [inline apinetlet/controllers/network_peering_controller.go:184]: is there a reason why this is loglevel=2? I would consider this as loglevel=1

File: apinetlet/controllers/network_peering_controller.go [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+			return networkingv1alpha1.NetworkPeeringClaimRef{}, "", "", nil
+		}
+		return peeringClaimRef, peering.Name, apinetNetwork.Spec.ID, nil
+	}
+
+	log.V(1).Info("No matching target peering found")
+	return networkingv1alpha1.NetworkPeeringClaimRef{}, "", "", nil
+}
+
+func (r *NetworkPeeringReconciler) enqueueNetworksHavingPeerings() handler.EventHandler {
```
Comment (balpert89) [inline apinetlet/controllers/network_peering_controller.go]: rename the func as this confuses

File: apinetlet/controllers/network_peering_controller.go (hunk 2 above)
Comment (kasabe28) [reply apinetlet/controllers/network_peering_controller.go]: Done

File: apinetlet/controllers/network_peering_controller.go (hunk 3 above)
Comment (kasabe28) [reply apinetlet/controllers/network_peering_controller.go]: Done

File: metalnetlet/controllers/network_controller.go [hunk 5]
```diff
@@ -144,6 +144,16 @@ func (r *NetworkReconciler) reconcile(ctx context.Context, log logr.Logger, netw
 			ID: vni,
 		},
 	}
+
+	for _, peeredId := range network.Spec.PeeredIDs {
+		id, err := networkid.ParseVNI(peeredId)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("error converting network peering ID to int32: %w", err)
```
Comment (afritzler) [inline metalnetlet/controllers/network_controller.go:151]:
```suggestion
			return ctrl.Result{}, fmt.Errorf("failed to parse peered network ID: %w", err)
```

**Review by afritzler — CHANGES_REQUESTED** [review]

**Comment by afritzler** [comment]
As https://github.com/ironcore-dev/ironcore/pull/1026 has been merged now, we can bump the `ironcore` dep and adjust the status of the network peering.

File: metalnetlet/controllers/network_controller.go (hunk 5 above)
Comment (kasabe28) [reply metalnetlet/controllers/network_controller.go]: Done

File: metalnetlet/controllers/network_controller.go [hunk 6]
```diff
@@ -144,6 +144,16 @@ func (r *NetworkReconciler) reconcile(ctx context.Context, log logr.Logger, netw
 			ID: vni,
 		},
 	}
+
```
Comment (afritzler) [inline metalnetlet/controllers/network_controller.go:147]: I think here we are missing the `Applied` state update of the `Network` peering status. `apinetlet` is patching the `Pending` state once peering refs are set. And once the `metalnetlet` is patching the corresponding `metalnet` `Network` resource it should patch the status to `Applied`.

**Review by afritzler — CHANGES_REQUESTED** [review]

File: metalnetlet/controllers/network_controller.go (hunk 6 above)
Comment (kasabe28) [reply metalnetlet/controllers/network_controller.go:147]: I think we need to wait for that to add `NetworkPeeringStatus` fields in `metalnet`

**Comment by balpert89** [comment]
currently blocked by https://github.com/ironcore-dev/metalnet/issues/189

**Review by afritzler — APPROVED** [review]

**Comment by afritzler** [comment]
@kasabe28 the PR looks good. Can you please rebase the PR branch?

**Pull request merged**